src/cordic.rs

```diff
@@ -0,0 +1,845 @@
+        (SinM<N0>, sine, Two, One, start(angle, modulus)),
+        (SinCosM<N0>, sine, Two, Two, start(angle, modulus)),
+        (ATan2<N0>, phase, Two, One, start(x, y)),
+        (Magnitude<N0>, modulus, Two, One, start(x, y)),
+        (ATan2Magnitude<N0>, phase, Two, Two, start(x, y)),
+        (CosH<N1>, hyperbolic_cosine, One, One, start(x)),
+        (SinH<N1>, hyperbolic_sine, One, One, start(x)),
+        (SinHCosH<N1>, hyperbolic_cosine, One, Two, start(x)),
+        (ATanH<N1>, arctanh, One, One, start(x)),
+    }
+
+    impls_multi_scale! {
+        (ATan<N0, N1, N2, N3, N4, N5, N6, N7>, arctangent, One, One, start(x)),
+        (Ln<N1, N2, N3, N4>, natural_logarithm, One, One, start(x)),
+        (Sqrt<N0, N1, N2>, square_root, One, One, start(x)),
+    }
+}
+
+/// Traits and structures related to precision type-states.
+pub mod prec {
+    /// Trait for precision type-states.
+    pub(crate) trait State {
+        /// Bit representation of the precision.
+        const BITS: u8;
+
+        /// Configure the resource to be represented
+        /// by this type-state.
+        fn set(w: &mut crate::stm32::cordic::csr::W);
+    }
+
+    /// 4 iterations.
+    pub struct P4;
+    /// 8 iterations.
+    pub struct P8;
+    /// 12 iterations.
+    pub struct P12;
+    /// 16 iterations.
+    pub struct P16;
+    /// 20 iterations.
+    pub struct P20;
+    /// 24 iterations.
+    pub struct P24;
+    /// 28 iterations.
+    pub struct P28;
+    /// 32 iterations.
+    pub struct P32;
+    /// 36 iterations.
+    pub struct P36;
+    /// 40 iterations.
+    pub struct P40;
+    /// 44 iterations.
+    pub struct P44;
+    /// 48 iterations.
+    pub struct P48;
+    /// 52 iterations.
+    pub struct P52;
+    /// 56 iterations.
+    pub struct P56;
+    /// 60 iterations.
+    pub struct P60;
+
+    macro_rules! impls {
+        ( $( ($NAME:ident, $BITS:expr) $(,)? )+ ) => {
+            $(
+                impl State for $NAME {
+                    const BITS: u8 = $BITS;
+
+                    #[inline]
+                    fn set(w: &mut crate::stm32::cordic::csr::W) {
+                        // SAFETY: reliant on valid type-state
+                        // implementations.
+                        unsafe { w.precision().bits(<Self as State>::BITS) };
+                    }
+                }
+            )+
+        };
+    }
+
+    impls! {
+        (P4, 1),
+        (P8, 2),
+        (P12, 3),
+        (P16, 4),
+        (P20, 5),
+        (P24, 6),
+        (P28, 7),
+        (P32, 8),
+        (P36, 9),
+        (P40, 10),
+        (P44, 11),
+        (P48, 12),
+        (P52, 13),
+        (P56, 14),
+        (P60, 15),
+    }
+}
+
+/// Cordic co-processor interface.
+#[allow(private_bounds)]
+pub struct Cordic<Arg, Res, Func, Prec>
+where
+    Arg: data_type::arg::State,
+    Res: data_type::res::State,
+    Func: func::State<Arg, Res>,
+    Prec: prec::State,
+{
+    rb: CORDIC,
```

Comment:
Just a thought, what if `rb` was `impl BorrowMut`? Then the user could

```rust
type CordicForSin = Cordic<_, _, Sin, _>;
type CordicForSqrt = Cordic<_, _, Sqrt, _>;

fn do_math_stuff(rb: &mut CORDIC) {
    {
        let mut cordic = rb.freeze_not_consuming();
        cordic.start(_);
        let _ = cordic.result();
    } // <--- rb no longer borrowed by cordic, it is free to be reused

    let mut cordic: CordicForSqrt = rb.freeze();
    
    cordic.start(_);
    let _ = cordic.result();
}
```

again, just a thought. Depending how the cordic is typically used, maybe not worth the extra generic arg.

Reply:
So, you nerdsniped me and I've been mulling this over since yesterday.

I just refactored the whole thing to be much stronger in terms of type relationships and enforcements.

As a result, now, adding this dynamic behavior is easy!

```rust
/// Traits and structures for dynamic function operation.
pub mod dynamic {
    use super::{prec, signature, types, Cordic, State};

    /// Any function can be invoked with this type-state.
    pub struct Any;

    /// A Cordic in dynamic mode.
    pub trait Mode<Arg, Res>
    where
        Arg: types::arg::State,
        Res: types::res::State,
    {
        /// Run a function with provided arguments and get the result.
        ///
        /// *Note: This employs the polling strategy.
        /// For less overhead, use static operations.*
        fn run<Func>(&mut self, args: Func::Arguments) -> Func::Results
        where
            Func: State<Arg, Res>;
    }

    impl<Arg, Res, Prec> Mode<Arg, Res> for Cordic<Arg, Res, Any, Prec>
    where
        Arg: types::arg::State,
        Res: types::res::State,
        Prec: prec::State,
    {
        fn run<Func>(&mut self, args: Func::Arguments) -> Func::Results
        where
            Func: State<Arg, Res>,
        {
            use signature::Property as _;

            self.apply_config::<Arg, Res, Func, Prec>();

            args.write(&self.rb.wdata);
            self.when_ready(|cordic| Func::Results::read(&cordic.rb.rdata))
        }
    }
}
```

So your example would look like this:

```rust
fn do_math_stuff(cordic: &mut Cordic<_, _, Any, _>) {
    let sin = cordic.run::<Sin>(_);
    let sqrt = cordic.run::<Sqrt<N0>>(_);
}
```

Reply:
Wow :)